sql/alps_submitter.go

```diff
@@ -329,30 +381,185 @@ func generateEstimatorCreator(estimator string, attrs []*attribute) (string, err
+		return e
+	}
+	if e = alpsTemplate.Execute(w, r); e != nil {
+		return fmt.Errorf("genALPS: failed executing template: %v", e)
+	}
 	return nil
 }
 
-func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB) error {
-	// TODO(uuleon): codegen & exec alps
+func trainALPS(wr *PipeWriter, pr *extendedSelect, cwd string) error {
+	var program bytes.Buffer
+	if e := genALPS(&program, pr); e != nil {
+		return e
+	}
+
+	code := program.String()
+
+	cw := &logChanWriter{wr: wr}
+	cmd := tensorflowCmd(cwd)
+	cmd.Stdin = &program
+	cmd.Stdout = cw
+	cmd.Stderr = cw
+	if e := cmd.Run(); e != nil {
+		return fmt.Errorf("code %v training failed %v", code, e)
+	}
 	return nil
 }
 
+func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB, cwd string) error {
+	if pr.train {
+		return trainALPS(w, pr, cwd)
+	} else {
+		return fmt.Errorf("inference not supported yet in ALPS")
+	}
+}
+
 const alpsTemplateText = `
 # coding: utf-8
+# Copyright (c) Antfin, Inc. All rights reserved.
 
-import tensorflow as tf
-
-from alps.feature import FeatureColumnsBuilder
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
 
+import os
 
-class SQLFlowFCBuilder(FeatureColumnsBuilder):
-	def build_feature_columns():
-		return {{.FeatureColumnsCode}}
+import tensorflow as tf
 
-# TODO(uuleon) needs FeatureSpecs and ALPS client
+from alps.client.base import submit_experiment, run_experiment
+from alps.conf.engine import LocalEngine
+from alps.estimator.column import SparseColumn, DenseColumn, RawColumn
+from alps.estimator.experiment import Experiment, TrainConf, EvalConf, EstimatorBuilder
+from alps.estimator.feature_column import embedding_column
+from alps.io.alps_dataset import AlpsDataset
+from alps.io.odps_io import OdpsConf
+from alps.io.reader.pyodps_reader import OdpsReader
+
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'    # for debug usage.
+tf.logging.set_verbosity(tf.logging.INFO)
+
+
+class SQLFlowEstimatorBuilder(EstimatorBuilder):
+    def build(self, experiment):
+        feature_columns = []
+		{{if .FeatureColumnCode}}
+        feature_columns.extend({{.FeatureColumnCode}})
+		{{end}}
+
+        return {{.EstimatorCreatorCode}}
+
+
+if __name__ == "__main__":	
+
+    ds = AlpsDataset(
+        num_epochs={{.DatasetConf.epoch}},
+		batch_size={{.DatasetConf.batch_size}},
+        reader=OdpsReader(
+            odps=OdpsConf(
+				accessid={{.OdpsConf.accessid}},
+				accesskey={{.OdpsConf.accesskey}},
+				endpoint={{.OdpsConf.endpoint}}
+			),
+            project={{.OdpsConf.project}},
+            table="{{.TableName}}",
+            field_names={{.Fields}},
+            features={{.X}},
+            labels={{.Y}}
+        )
+    )
+
+    experiment = Experiment(
+        user="sqlflow",
+		engine=LocalEngine(),
+		{{if .TrainSpec.max_steps}}
+        train=TrainConf(input=ds, max_steps={{.TrainSpec.max_steps}}),
+		{{else}}
+		train=TrainConf(input=ds),
+		{{end}}
+        scratch_dir="./scratch",
```

Comment:
Is the `scratch_dir` the current working directory (`cwd`)? 

If so, it seems that the function `genALPS` should take an addition parameter `cwd`  and uses the value to replace the value of `scratch_dir` here.

If not, please make sure that multiple locally-running Alps jobs spawned from the same SQLFlow server process would not write to the same `./scratch` directory.

Reply:
`scratch_dir` is the directory where `checkpoint` and `model` are stored. Fixed directories are really inappropriate. Thank you for pointing out.